runtime/to-json-processor.h

```diff
@@ -0,0 +1,167 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2020 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include "runtime/kphp_core.h"
+#include "runtime/json-writer.h"
+#include "runtime/json-processor-utils.h"
+
+template <class Tag>
+class ToJsonVisitor {
+public:
+  explicit ToJsonVisitor(impl_::JsonWriter &writer) noexcept
+    : writer_(writer) {}
+
+  template<class T>
+  void operator()(std::string_view key, const T &value, bool array_as_hashmap = false) noexcept {
+    writer_.Key(key);
+    process_impl(value, array_as_hashmap);
+  }
+
+  void set_precision(size_t precision) noexcept {
+    writer_.SetDoublePrecision(precision);
+  }
+
+  void restore_precision() noexcept {
+    writer_.SetDoublePrecision(0);
+  }
+
+private:
+  template<class T>
+  void process_impl(const T &value, bool /*array_as_hashmap*/) noexcept {
+    add_value(value);
+  }
+
+  template<class T>
+  void process_impl(const Optional<T> &value, bool array_as_hashmap) noexcept {
+    auto process_impl_lambda = [this, array_as_hashmap](const auto &val) { return this->process_impl(val, array_as_hashmap); };
+    call_fun_on_optional_value(process_impl_lambda, value);
+  }
+
+  template<class I>
+  void process_impl(const class_instance<I> &instance, bool /*array_as_hashmap*/) noexcept;
+
+  template<class T>
+  void process_vector(const array<T> &array) noexcept {
+    writer_.StartArray();
+    for (const auto &elem : array) {
+      process_impl(elem.get_value(), false);
+    }
+    writer_.EndArray();
+  }
+
+  template<class T>
+  void process_map(const array<T> &array) noexcept {
+    writer_.StartObject();
+    for (const auto &elem : array) {
+      const auto &key = elem.get_key().to_string();
+      (*this)({key.c_str(), key.size()}, elem.get_value());
+    }
+    writer_.EndObject();
+  }
+
+  template<class T>
+  void process_impl(const array<T> &array, bool array_as_hashmap) noexcept {
+    (array.is_vector() || array.is_pseudo_vector()) && !array_as_hashmap ? process_vector(array) : process_map(array);
+  }
+
+  // support of array<Unknown> compilation
+  void process_impl(const Unknown &/*elem*/, bool /*array_as_hashmap*/) noexcept {}
```

Comment:
Could you elaborate why we need to support unknown types compilation?
AFAIK, unknown types should be catched during the type checking, they should not make it till the C++ code compilation. Or even if they do, there is nothing else than failing the compilation that can make sense in my opinion.

Is there some specific case we need to support here?

Reply:
This code intended for deal with empty arrays:
```
class A {
  public $arr = [];
}
```
In cpp `arr` transpiled to `array<Unknown>`. So such overload allows to avoid ruin of compilation.